apps/src/templates/projects/PersonalProjectsPublishedCell.jsx

```diff
@@ -16,13 +17,32 @@ class PersonalProjectsPublishedCell extends Component {
     publishMethod: PropTypes.oneOf([publishMethods.CHEVRON, publishMethods.BUTTON]).isRequired,
     showPublishDialog: PropTypes.func.isRequired,
     unpublishProject: PropTypes.func.isRequired,
+    userId: PropTypes.string,
   };
 
   onPublish = () => {
+    firehoseClient.putRecord(
+      {
+        study: 'project-publish',
+        study_group: 'publish-button',
+        event: 'publish',
+        user_id: this.props.userId,
```

Comment:
I haven't done this kind of logging yet. 1.) can I just add in extra fields like this? The examples I've found (mostly #22486) don't. 2.) Is there a way to check locally that this logging is working or do I just ask Poorva to check Redshift once the change hits staging?

Reply:
1.) So this should actually work, but it was probably not meant to work this way.

You can't normally just add new fields like this, you would first need to add columns to the table that this logs to. The available fields are here:
https://github.com/code-dot-org/code-dot-org/blob/99db4d453e4007150d82d960968f835d9913f5ad/apps/src/lib/util/firehose.js#L10-L20

However `user_id` actually exists already, but it's not meant to be specified in the data argument to the `putRecord` call. It's a field that is automatically added to the firehose record if you include `includeUserId: true` in the options argument to `putRecord`: https://github.com/code-dot-org/code-dot-org/blob/99db4d453e4007150d82d960968f835d9913f5ad/apps/src/lib/util/firehose.js#L183

Unfortunately, you then run into the problem that the `pageConstants` reducer isn't used on non-puzzle pages, so the firehose code can't actually look up the user_id when you're in the project gallery:
https://github.com/code-dot-org/code-dot-org/blob/99db4d453e4007150d82d960968f835d9913f5ad/apps/src/lib/util/firehose.js#L159-L162

The way that `addCommonValues` is written though, your code should actually just work. It won't touch your manually specified `user_id`, and firehose/redshift won't know the difference between this and a normal automatically populated `user_id`.

I think you have two options:
1. Add the pageConstants reducer to the redux store in the project gallery, and populate the user_id so that the `includeUserId` option works as intended.
2. Update the usage comment for `putRecord` to officially allow a `user_id` property (but also mention the `includeUserId` option in the comment), and update the code above so that it doesn't override a manually specified `user_id` with the one from redux. You probably want to do this.

Reply:
2.) I think that you can pass the `alwaysPut` option to `putRecord` to force it to log to firehose in development mode, but you'd still have to ask Poorva to check redshift to see if the record was added (or set up your own access to redshift). If my memory serves, you also have to wait a day or so for the record to actually show up.

I usually don't bother with that. In development mode `putRecord` logs to the JS console instead of to firehose, I just make sure it does that.